bpo-38364: unwrap partialmethods just like we unwrap partials

mjpieters opened the pull request:
The inspect.isgeneratorfunction, inspect.iscoroutinefunction and inspect.isasyncgenfunction already unwrap functools.partial objects, this patch adds support for partialmethod objects as well.

<!-- issue-number: [bpo-38364](https://bugs.python.org/issue38364) -->
https://bugs.python.org/issue38364
<!-- /issue-number -->

mjpieters committed:
📜🤖 Added by blurb_it.

mjpieters committed:
bpo-38364: unwrap partialmethods just like we unwrap partials

The inspect.isgeneratorfunction, inspect.iscoroutinefunction and inspect.isasyncgenfunction already unwrap functools.partial objects, this patch adds support for partialmethod objects as well.

taleinat commented on Lib/inspect.py:
```diff
@@ -170,8 +170,10 @@ def isfunction(object):
     """Return true if ``f`` is a function (or a method or functools.partial
-    wrapper wrapping a function) whose code object has the given ``flag``
+    wrapper wrapping a function or a functools.partialmethod wrapping a
+    function) whose code object has the given ``flag``
```
Shouldn't this be "or a functools.partialmethod wrapping a method"?

mjpieters replied:
What matters is the function underneath. The partialmethod object can wrap any descriptor type, but its the function that has the flags.

taleinat commented on Lib/test/test_inspect.py:
```diff
@@ -167,12 +167,33 @@ def test_iscoroutine(self):
         gen_coro = gen_coroutine_function_example(1)
         coro = coroutine_function_example(1)
 
+        class PMClass:
```
How about a test case for a partialmethod wrapping a normal method?

mjpieters replied:
Testing how inspect interacts with `partialmethod()` on regular methods is [covered elsewhere already](https://github.com/python/cpython/blob/9a669d58e8cb586fba38c84d5b631cd8a95d0c0c/Lib/test/test_inspect.py#L859-L871); these are the coroutine-specific tests.

kumaraditya303 reviewed (changes requested):
There are merge conflicts.

kumaraditya303 commented on Doc/library/inspect.rst:
```diff
@@ -354,6 +357,10 @@ attributes (see :ref:`import-mod-attrs` for module attributes):
       wrapped function is a :term:`coroutine function`.
 
+   .. versionchanged:: 3.12
+      Functions wrapped in :func:`functools.partialpartial` now return ``True``
```
```suggestion
      Functions wrapped in :func:`functools.partialmethod` now return ``True``
```

mjpieters replied:
🤦‍♂️ Good catch, that's a really dumb merge error on my part!

kumaraditya303 commented on Doc/library/inspect.rst:
```diff
@@ -418,6 +425,9 @@ attributes (see :ref:`import-mod-attrs` for module attributes):
       wrapped function is a :term:`asynchronous generator` function.
 
+   .. versionchanged:: 3.12
+      Functions wrapped in :func:`functools.partialpartial` now return ``True``
```
```suggestion
      Functions wrapped in :func:`functools.partialmethod` now return ``True``
```

kumaraditya303 commented on Lib/functools.py:
```diff
@@ -423,6 +423,17 @@ def _unwrap_partial(func):
+    prev = None
+    while func is not prev:
+        prev = func
+        while isinstance(getattr(func, "_partialmethod", None), partialmethod):
```
Why getattr? What objects have `_partialmethod`?

mjpieters replied:
`partialmethod(...).__get__()` can return a closure with `_partialmethod` pointing to the instance. This happens if the wrapped callable is not a descriptor, or if the descriptor returned itself again.

See the [`partialmethod._make_unbound_method()` implementation](https://github.com/python/cpython/blob/98308dbeb110198ebe28bdb7720d3671b3e7f57b/Lib/functools.py#L385-L391).

`getattr()` is used because the attribute is optional, and you'd not want this to break even if the supplied object does have a `_partialmethod` attribute that is not actually a `partialmethod` instance.

encukou committed:
Merge in the main branch

encukou committed:
Update versionchanged directives

encukou committed:
Rename _partialmethod to __partialmethod__

Since we'rs checking this attribute on arbitrary function-like objects,
we should use the namespace reserved for core Python.
